### PR TITLE
Fix dependencies type changes scenario in project model server.

### PR DESCRIPTION
**Pull request by troydai**
Issue: #2995

### CODE REVIEW

**Comment by troydai** [comment]
@dotnet-bot test  Ubuntu x64 Release Build  please.

**Comment by troydai** [comment]
@dotnet-bot test RHEL7.2 x64 Release Build please

**Comment by troydai** [comment]
I'm investigating the CI test failure

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

File: test/dotnet-projectmodel-server.Tests/DthTests.cs [hunk 1]
```diff
@@ -271,17 +271,71 @@ public void DthDependencies_UpdateGlobalJson_RefreshDependencies()
...
+                      .RetrievePayloadAs<JObject>()
+                      .AssertProperty<JArray>("Errors", array => array.Count == 0)
+                      .AssertProperty<JArray>("Warnings", array => array.Count == 0);
+
+                // Overwrite the global.json to add search path to ext back
+                File.WriteAllText(
+                    Path.Combine(projectPath, "home", GlobalSettings.FileName),
+                    JsonConvert.SerializeObject(new { projects = new string[] { "src", "../ext" } }));
+
+                // Console.WriteLine($"attach to {System.Diagnostics.Process.GetCurrentProcess().Id}");
```
Comment (troydai) [inline test/dotnet-projectmodel-server.Tests/DthTests.cs]: will clean this up.

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please
@dotnet-bot test RHEL7.2 x64 Release Build please

**Comment by troydai** [comment]
****

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please****

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet-bot test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
@dotnet test Ubuntu x64 Release Build please

**Comment by troydai** [comment]
The failure on Ubuntu isn't related to changes in this PR

```
11:38:53 [39;49mBuild failed: System.Exception: Unknown channel - rel-1.0.0
11:38:53    at Microsoft.DotNet.Cli.Build.Monikers.GetSdkDebianPackageName(BuildTargetContext c)
11:38:53    at Microsoft.DotNet.Cli.Build.DebTargets.GenerateSdkDeb(BuildTargetContext c)
11:38:53    at Microsoft.DotNet.Cli.Build.Framework.BuildContext.ExecTarget(BuildTarget target)
11:38:56 
```

Any idea?

**Comment by troydai** [comment]
Base on top of the `rel/1.0.0`

**Comment by piotrpMSFT** [comment]
This looks ok to me overall. I'm not super familiar with the DTH codebase however.

**Comment by troydai** [comment]
@eerhardt mind take a look as well? I want to merge this first.

File: TestAssets/ProjectModelServer/DthUpdateSearchPathSample/home/src/MainProject/project.json [hunk 2]
```diff
@@ -1,8 +1,10 @@
 {
   "dependencies": {
-    "Newtonsoft.Json": "6.0.8"
+    "Newtonsoft.Json": "8.0.3"
```
Comment (eerhardt) [inline TestAssets/ProjectModelServer/DthUpdateSearchPathSample/home/src/MainProject/project.json]: FYI - if you update this to "9.0.1" you should need any `imports` in your netstandard1.5 AFAIK.

**Comment by eerhardt** [comment]
>             snapshot.ProjectSearchPaths = currentSearchPaths.ToList();

Does this need to be updated to `projectSearchPaths`?  I'm a bit confused at the naming since the parameter is named "projectSearchPaths" and this property is named `snapshot.ProjectSearchPaths`, but the variable is `currentSearchPaths`.  They all sound related, but the naming doesn't make it clear what the difference is.

---
Refers to: src/dotnet/commands/dotnet-projectmodel-server/InternalModels/ProjectSnapshot.cs:35 in eb03a74. [](commit_id = eb03a745bc6123822ad7681fc3626ea2b8107c2c, deletion_comment = False)

File: TestAssets/ProjectModelServer/DthUpdateSearchPathSample/home/src/MainProject/project.json (hunk 2 above)
Comment (troydai) [inline TestAssets/ProjectModelServer/DthUpdateSearchPathSample/home/src/MainProject/project.json]: will do.

**Comment by eerhardt** [comment]
Looks good.  Just a couple minor comments.

**Comment by troydai** [comment]
they are the same thing, i can change the variable name to make it clear if you insist.

**Comment by eerhardt** [comment]
> they are the same thing, i can change the variable name to make it clear if you insist.

I really think it would help understand the difference.  It is really odd when the method is:

```
var myProjectSearchPaths = foo;
var snapshot = ProjectSnapshot.Create(bar, baz, projectSearchPaths: myProjectSearchPaths);

snapshot.ProjectSearchPaths == myProjectSearchPaths  ?     <=== this isn't necessarily true
```

**Comment by troydai** [comment]
@eerhardt what's your real concerns here? I can't really grasp it. Is it he name make the logic difficult to understand or you're wondering what's the property and variable for?

**Comment by eerhardt** [comment]
:shipit:

**Comment by troydai** [comment]
An unrelated publish test failed on Windows_NT x86
```
Regression

Microsoft.DotNet.Tools.Publish.Tests.PublishPortableTests.PortableAppWithRuntimeTargetsIsPublishedCorrectly (from (empty))

Failing for the past 1 build (Since Failed#1827 )
Took 2.5 sec.
Stacktrace

MESSAGE:
File "System.Linq.dll" should not be found in directory "D:\j\workspace\debug_windows---54dcb9ec\test\dotnet-publish.Tests\bin\Debug\netcoreapp1.0\PortableAppWithRuntimeTargetsIsPublishedCorrectly\PortableTests\PortableAppWithNative\bin\Debug\netcoreapp1.0\publish".
+++++++++++++++++++
STACK TRACE:
at FluentAssertions.Execution.LateBoundTestFramework.Throw(String message) in D:\Workspaces\FluentAssertions\Src\Shared\Execution\LateBoundTestFramework.cs:line 23 at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message) in D:\Workspaces\FluentAssertions\Src\FluentAssertions.Dotnet\Execution\TestFrameworkProvider.cs:line 38 at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message) in D:\Workspaces\FluentAssertions\Src\Core\Execution\DefaultAssertionStrategy.cs:line 25 at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args) in D:\Workspaces\FluentAssertions\Src\Core\Execution\AssertionScope.cs:line 197 at Microsoft.DotNet.Tools.Test.Utilities.DirectoryInfoAssertions.NotHaveFile(String expectedFile) in D:\j\workspace\debug_windows---54dcb9ec\test\Microsoft.DotNet.Tools.Tests.Utilities\Assertions\DirectoryInfoAssertions.cs:line 44 at Microsoft.DotNet.Tools.Publish.Tests.PublishPortableTests.PortableAppWithRuntimeTargetsIsPublishedCorrectly() in D:\j\workspace\debug_windows---54dcb9ec\test\dotnet-publish.Tests\PublishPortableTests.cs:line 37
```

@dotnet-bot test Windows_NT x86 Debug Build please.

**Comment by troydai** [comment]
OSX failed for unrelated reason:
```
java.io.IOException: remote file operation failed: /Users/dotnet-bot/j/workspace/dotnet_cli/rel_1.0.0/release_osx_x64_prtest at hudson.remoting.Channel@447b0121:dci-macpro-12: hudson.util.IOException2: There are some problems during the conversion into JUnit reports: 
```

@dotnet-bot test OSX x64 Release Build please.

**Comment by troydai** [comment]
@dotnet-bot test OSX x64 Release Build please

**Comment by troydai** [comment]
@brthor is this OSX failure known? http://dotnet-ci.cloudapp.net/job/dotnet_cli/job/rel_1.0.0/job/release_osx_x64_prtest/2294/console

**Comment by brthor** [comment]
cc @mmitche `07:24:10 rm: /Users/dotnet-bot/.local/share/NuGet/Cache: Permission denied`
on dci-macpro-04

test OSX x64 Release Build please

**Comment by mmitche** [comment]
@brthor Let's pop on a machine and check the permissions on that directory.

**Comment by mmitche** [comment]
@brthor When did this change go in?  What happens if two builds are running in parallel?  Can we simply put the machine cache elsewhere?

**Comment by mmitche** [comment]
@brthor It was created as root.  I deleted it, but I'm worried about the implications of that machine cache removal.

**Comment by brthor** [comment]
Which change @mmitche ?

We put our cache directly under the repo currently to avoid that problem. 

@emgarten @joelverhagen Maybe this is the new fallback location?

**Comment by mmitche** [comment]
@brthor Why are we trying to remove the .local cache then?

(also I accidentally hit close and reopen)

**Comment by brthor** [comment]
Ah @ericstj Added this to prevent getting outdated `1.0.0` packages as we rebuilt them for rtm. @ericstj Is there any way to avoid this?

**Comment by ericstj** [comment]
That isn't the same .local/share/NuGet/Cache is the NuGet HTTP cache.  It should be removable by a regular user.

**Comment by mmitche** [comment]
It was created by root.  It could be leftovers from when the jenkins process was running as root.  We can just remove them, though I thought I did before.

I'm primarily worried about the implications of this + multi-threading.  What happens if we're restoring two sets of packages at the same time in different jobs?

**Comment by ericstj** [comment]
NuGet already doesn't handle multiple restores running concurrently and has punted on a lot of the bugs we've filed so I wouldn't recommend doing that.

**Comment by mmitche** [comment]
@ericstj We don't have the resources to restrict the macs to a single build at a time.  I haven't seen any restore bugs related to this in a long time.

**Comment by ericstj** [comment]
@mmitche probably getting lucky.  The sorts of problems we've hit are 1) too many requests sent to single server 2) hangs.  Both of these had to deal with lack of cross-process communication between instances of nuget.

An alternate to deleting the directory is running with `--no-cache` but it needs to be added to every place we call restore.

**Comment by mmitche** [comment]
@ericstj Maybe lucky?  Not sure.  Anyways, it also might be that nuget has some built in fallback if the cache can't be written to (because there was no way for it to write to the cache for the last couple weeks, given permissions)

Anyways, we should probably just leave it multi-threaded for now, since we really cannot make it single threaded.  Alternatively maybe we create the nuget directory as non-writable....

What is the .NUget directory for?

**Comment by troydai** [comment]
oh, excellent. time out error backs agin.

@dotnet-bot test WIndows_NT x64 Release Build please

**Comment by ericstj** [comment]
This particular directory we are deleting is the HTTP cache.  When nuget makes a request to an HTTP source it stores results here.   You can think of it like temporary internet files.

**Comment by mmitche** [comment]
@troydai timeout failures?  Does that manifest as:

error NU1010: The type of dependency Newtonsoft.Json was changed.",
Indication 1

Is there a better string in that log file that could indicate timeout?

**Comment by troydai** [comment]
Some of the DthTests are known to be vulnerable if the test machine's performance is poor. Usually it recovers in rerun.

**Comment by troydai** [comment]
@dotnet-bot test OpenSUSE13.2 x64 Debug Build please.

**Comment by mmitche** [comment]
@troydai We should fix the tests.  This is a waste of resources to keep retrying

**Comment by mmitche** [comment]
If we don't have the resources to fix them now, can you point me to the issues that have been filed?

**Comment by troydai** [comment]
@mmitche they just shows up these two days again. We don't have issue tracking them.

**Comment by mmitche** [comment]
@troydai It's good hygiene to keep track of these, can you file?  Keeping our tests in a good state keeps us out of a state where we think "well that just doesn't work, retry".

**Comment by troydai** [comment]
@mmitche https://github.com/dotnet/cli/issues/3614 can you help me assign proper milestone and labels?

**Comment by mmitche** [comment]
@treoy I don't think I'm familiar enough to make a call on timeframe. @piotrpMSFT can you help with that?

**Pull request merged**